### ISSUE TITLE
Inconsistent computations

### ISSUE THREAD

**Issue by sw1sh** [plaidml/plaidml#47]
Network computes different numbers than tensorflow backend does. Comparing the result of just the first convolutional layer of VGG19, the mean squared error between outputs of both backends is 
 of order 10^-11, second 10^-9, and after pooling it's already 36(!), which totally screws with a prediction.
This [lion picure](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1e/Cecil_the_lion_at_Hwange_National_Park_%284516560206%29.jpg/330px-Cecil_the_lion_at_Hwange_National_Park_%284516560206%29.jpg) incorrectly predicts `ocarina`, while tensorflow backend predicting `lion` correctly.
 
```python
import numpy as np
from urllib.request import urlopen

import plaidml.keras
plaidml.keras.install_backend()

import keras
from keras.applications import VGG19
from keras.applications.vgg19 import preprocess_input, decode_predictions
from keras.models import Model
from keras.preprocessing import image

url = 'https://upload.wikimedia.org/wikipedia/commons/thumb/1/1e/Cecil_the_lion_at_Hwange_National_Park_%284516560206%29.jpg/330px-Cecil_the_lion_at_Hwange_National_Park_%284516560206%29.jpg'

img = image.pil_image.open(urlopen(url)).resize((224, 224))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

model = VGG19()
decode_predictions(model.predict(x, batch_size=1))
```

**Comment by sw1sh** [comment]
Actually if I just run the last line `decode_predictions(model.predict(x, batch_size=1))` multiple times it produces different results every time! Sometimes it indeed predicts `lion` correctly but probabilities are always different, and other times it is totally wrong.

**Comment by brianretford** [comment]
I can't reproduce on linux / AMD fiji. Can you provide a dump of uname -a and clinfo for the card you're running?

**Comment by sw1sh** [comment]
Trying on Mac:
```
-> % uname -a
Darwin ibmmac 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64
-> % clinfo
Number of platforms                               1
  Platform Name                                   Apple
  Platform Vendor                                 Apple
  Platform Version                                OpenCL 1.2 (May 26 2017 12:59:48)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event

  Platform Name                                   Apple
Number of devices                                 3
  Device Name                                     Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  Device Vendor                                   Intel
  Device Vendor ID                                0xffffffff
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.1
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     CPU
  Device Available                                Yes
  Device Profile                                  FULL_PROFILE
  Max compute units                               8
  Max clock frequency                             2500MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             1024x1x1
  Max work group size                             1024
  Compiler Available                              Yes
  Linker Available                                Yes
  Preferred work group size multiple              1
  Preferred / native vector sizes
    char                                                16 / 16
    short                                                8 / 8
    int                                                  4 / 4
    long                                                 2 / 2
    half                                                 0 / 0        (n/a)
    float                                                4 / 4
    double                                               2 / 2        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    64, Little-Endian
  Global memory size                              17179869184 (16GiB)
  Error Correction support                        No
  Max memory allocation                           4294967296 (4GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        64
  Global Memory cache line size                   6291456 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   1 bytes
    Pitch alignment for 2D image buffers          1 bytes
    Max 2D image size                             8192x8192 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Global
  Local memory size                               32768 (32KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     4096 (4KiB)
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      1ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_APPLE_fp64_basic_ops cl_APPLE_fixed_alpha_channel_orders cl_APPLE_biased_fixed_point_image_formats cl_APPLE_command_queue_priority

  Device Name                                     Iris Pro
  Device Vendor                                   Intel
  Device Vendor ID                                0x1024500
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.2(Jun 15 2017 18:45:12)
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     GPU
  Device Available                                Yes
  Device Profile                                  FULL_PROFILE
  Max compute units                               40
  Max clock frequency                             1200MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             512x512x512
  Max work group size                             512
  Compiler Available                              Yes
  Linker Available                                Yes
  Preferred work group size multiple              32
  Preferred / native vector sizes
    char                                                 1 / 1
    short                                                1 / 1
    int                                                  1 / 1
    long                                                 1 / 1
    half                                                 0 / 0        (n/a)
    float                                                1 / 1
    double                                               0 / 0        (n/a)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (n/a)
  Address bits                                    64, Little-Endian
  Global memory size                              1610612736 (1.5GiB)
  Error Correction support                        No
  Max memory allocation                           402653184 (384MiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            25165824 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   4 bytes
    Pitch alignment for 2D image buffers          32 bytes
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      80ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_image2d_from_buffer cl_khr_gl_depth_images cl_khr_depth_images cl_khr_3d_image_writes

  Device Name                                     AMD Radeon R9 M370X Compute Engine
  Device Vendor                                   AMD
  Device Vendor ID                                0x1021c00
  Device Version                                  OpenCL 1.2
  Driver Version                                  1.2 (Jun 15 2017 18:41:53)
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     GPU
  Device Available                                Yes
  Device Profile                                  FULL_PROFILE
  Max compute units                               10
  Max clock frequency                             800MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Compiler Available                              Yes
  Linker Available                                Yes
  Preferred work group size multiple              64
  Preferred / native vector sizes
    char                                                 4 / 4
    short                                                2 / 2
    int                                                  1 / 1
    long                                                 1 / 1
    half                                                 0 / 0        (n/a)
    float                                                1 / 1
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    32, Little-Endian
  Global memory size                              2147483648 (2GiB)
  Error Correction support                        No
  Max memory allocation                           536870912 (512MiB)
  Unified memory for Host and Device              No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       32768 bits (4096 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   256 bytes
    Pitch alignment for 2D image buffers          256 bytes
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               32768 (32KiB)
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      37ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            134217728 (128MiB)
  Built-in kernels
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_image2d_from_buffer cl_khr_depth_images cl_APPLE_command_queue_priority cl_APPLE_command_queue_select_compute_units cl_khr_fp64

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Apple
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [P0]
  clCreateContext(NULL, ...) [default]            Success [P0]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (2)
    Platform Name                                 Apple
    Device Name                                   Iris Pro
    Device Name                                   AMD Radeon R9 M370X Compute Engine
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  Success (1)
    Platform Name                                 Apple
    Device Name                                   Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (2)
    Platform Name                                 Apple
    Device Name                                   Iris Pro
    Device Name                                   AMD Radeon R9 M370X Compute Engine
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  <checkNullCtxFromType:2437: create context from type CL_DEVICE_TYPE_CUSTOM : error -30>
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (3)
    Platform Name                                 Apple
    Device Name                                   Iris Pro
    Device Name                                   AMD Radeon R9 M370X Compute Engine
    Device Name                                   Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
	NOTE:	your OpenCL library only supports OpenCL 1.0,
		but some installed platforms support OpenCL 1.2.
		Programs using 1.2 features may crash
		or behave unexepectedly
```

**Comment by jbruestle** [comment]
Ran 10 times on MacOS 10.12.16.  It works 10 out of 10 on the Isis/Intel GPU, but on AMD (AMD Radeon R9 M370X Compute Engine) I get for 10 runs:

```
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
[[(u'n02129165', u'lion', 0.99999523), (u'n02112137', u'chow', 1.7141298e-06), (u'n02115913', u'dhole', 1.3489707e-06), (u'n02117135', u'hyena', 9.138181e-07), (u'n02410509', u'bison', 2.5300733e-07)]]
[[(u'n02129165', u'lion', 0.99999309), (u'n02112137', u'chow', 2.2982056e-06), (u'n02115913', u'dhole', 2.2519562e-06), (u'n02117135', u'hyena', 1.1406268e-06), (u'n02410509', u'bison', 3.5231079e-07)]]
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
[[(u'n02129165', u'lion', 0.99999738), (u'n02115913', u'dhole', 1.1977942e-06), (u'n02112137', u'chow', 5.2453163e-07), (u'n02117135', u'hyena', 4.734018e-07), (u'n02410509', u'bison', 2.0128073e-07)]]
[[(u'n02129165', u'lion', 0.99998963), (u'n02112137', u'chow', 4.9734986e-06), (u'n02115913', u'dhole', 1.921713e-06), (u'n02117135', u'hyena', 1.1732755e-06), (u'n02410509', u'bison', 9.1051476e-07)]]
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
[[(u'n04090263', u'rifle', 1.0), (u'n15075141', u'toilet_tissue', 0.0), (u'n02317335', u'starfish', 0.0), (u'n02391049', u'zebra', 0.0), (u'n02389026', u'sorrel', 0.0)]]
```

**Comment by jbruestle** [comment]
I have a repo here, so I'll start debugging it, but just for curiosity, does the Iris Pro in your machine work?  You seem to have the same machine as me.

**Comment by sw1sh** [comment]
Yes, it works on Iris Pro. Also run on cpu and got `'unable to run OpenCL kernel: CL_INVALID_WORK_GROUP_SIZE` error, but this is a whole other issue I guess.

**Comment by jbruestle** [comment]
Yes, the CPU issue is an known limitation of Apple's CPU OpenCL which allows only 1 thread per work-group.  We have not addressed this, as the resulting CPU performance is likely to be underwhelming.  As for the main issue, I've tracked the problem down to a set of kernels that behave incorrectly, but only on AMD and only on Apple's drivers.  I'm still checking, but the kernels strongly appear to be correct, and I believe this may be a bug in the OpenCL drivers themselves.  There is some discussion of similar correctness issues with Apple's OpenCL drivers in other OpenCL projects.  That said, it appears to happen only for a subset of kernels.  There is a configuration change that seems to work around the problem by avoiding such kernels.  There may be a performance penalty for the change, and since I don't know what the specific driver issue is, I'm not sure if this workaround will fully resolve all correctness issues.  However, it did fix the issue for me.  

Could you please test the following change and see if it works?  Specifically, I changed max_regs for the AMD in plaidml/experimental.json to 2048.  I've made a local build of the plaidml pip.  Make sure to uninstall the official build before installing this one.  The pip is available here: https://storage.googleapis.com/vertexai-releases/git-issue-47/plaidml-0.0.0.dev0-py2.py3-none-macosx_10_10_x86_64.whl

Longer term, we will try to make a simpler repro for the driver issue and submit to Apple, and also look into adding Metal support to the HAL, as I suspect the Metal drivers are more actively maintained, and we will need Metal support for FP16 anyway.

**Comment by sw1sh** [comment]
Yes, your build fixed this issue for me too, great job!

**Comment by jbruestle** [comment]
Pushed the fixes to master and made new pips (version 0.1.2).  Closing issue as fixed, reopen if the new pips don't work.

**Comment by jlussi** [comment]
> Yes, the CPU issue is an known limitation of Apple's CPU OpenCL which allows only 1 thread per work-group. We have not addressed this, as the resulting CPU performance is likely to be underwhelming. As for the main issue, I've tracked the problem down to a set of kernels that behave incorrectly, but only on AMD and only on Apple's drivers. I'm still checking, but the kernels strongly appear to be correct, and I believe this may be a bug in the OpenCL drivers themselves. There is some discussion of similar correctness issues with Apple's OpenCL drivers in other OpenCL projects. That said, it appears to happen only for a subset of kernels. There is a configuration change that seems to work around the problem by avoiding such kernels. There may be a performance penalty for the change, and since I don't know what the specific driver issue is, I'm not sure if this workaround will fully resolve all correctness issues. However, it did fix the issue for me.
> 
> Could you please test the following change and see if it works? Specifically, I changed max_regs for the AMD in plaidml/experimental.json to 2048. I've made a local build of the plaidml pip. Make sure to uninstall the official build before installing this one. The pip is available here: https://storage.googleapis.com/vertexai-releases/git-issue-47/plaidml-0.0.0.dev0-py2.py3-none-macosx_10_10_x86_64.whl
> 
> Longer term, we will try to make a simpler repro for the driver issue and submit to Apple, and also look into adding Metal support to the HAL, as I suspect the Metal drivers are more actively maintained, and we will need Metal support for FP16 anyway.

 Hey, 
Sorry I know the conversation is a while ago, but I am facing similar issues using the AMD VEGA 20 card. Although you implemented the fix, for me, the learning differs massively if I run it on CPU or GPU, mostly it doesn't converge properly with the GPU. Here for comparison:

GPU:
Epoch 1/40
281/281 [==============================] - 386s 1s/step - loss: 0.7484 - acc: 0.5923 - recall: 0.6787 - val_loss: 1.2225 - val_acc: 0.6412 - val_recall: 0.8492
Epoch 2/40
281/281 [==============================] - 350s 1s/step - loss: 0.6085 - acc: 0.6799 - recall: 0.7000 - val_loss: 1.1240 - val_acc: 0.5719 - val_recall: 0.8315
Epoch 3/40
281/281 [==============================] - 369s 1s/step - loss: 0.5976 - acc: 0.7000 - recall: 0.7457 - val_loss: 0.6783 - val_acc: 0.6927 - val_recall: 0.7176
Epoch 4/40
281/281 [==============================] - 390s 1s/step - loss: 0.5878 - acc: 0.7086 - recall: 0.7328 - val_loss: 0.6793 - val_acc: 0.6989 - val_recall: 0.8193
Epoch 5/40
281/281 [==============================] - 372s 1s/step - loss: 0.5761 - acc: 0.7075 - recall: 0.7304 - val_loss: 0.6274 - val_acc: 0.6927 - val_recall: 0.6125
Epoch 6/40
281/281 [==============================] - 381s 1s/step - loss: 0.5620 - acc: 0.7137 - recall: 0.7183 - val_loss: 0.6101 - val_acc: 0.6838 - val_recall: 0.5884
Epoch 7/40
281/281 [==============================] - 353s 1s/step - loss: 0.5483 - acc: 0.7295 - recall: 0.7281 - val_loss: 0.5534 - val_acc: 0.7211 - val_recall: 0.7350
Epoch 8/40
281/281 [==============================] - 452s 2s/step - loss: 0.5324 - acc: 0.7402 - recall: 0.7201 - val_loss: 0.5808 - val_acc: 0.7194 - val_recall: 0.7659



and on the other hand the CPU: 

Epoch 2/40
281/281 [==============================] - 143s 509ms/step - loss: 0.6035 - acc: 0.6775 - recall: 0.6793 - val_loss: 1.7410 - val_acc: 0.5835 - val_recall: 0.5224
Epoch 3/40
281/281 [==============================] - 143s 509ms/step - loss: 0.5970 - acc: 0.6810 - recall: 0.6979 - val_loss: 0.6803 - val_acc: 0.6208 - val_recall: 0.7823
Epoch 4/40
281/281 [==============================] - 143s 510ms/step - loss: 0.5844 - acc: 0.6942 - recall: 0.7146 - val_loss: 0.6377 - val_acc: 0.6243 - val_recall: 0.4733
Epoch 5/40
281/281 [==============================] - 143s 509ms/step - loss: 0.6111 - acc: 0.6819 - recall: 0.7422 - val_loss: 1.3407 - val_acc: 0.5950 - val_recall: 0.7550
Epoch 6/40
281/281 [==============================] - 143s 509ms/step - loss: 0.6037 - acc: 0.6966 - recall: 0.7658 - val_loss: 0.7204 - val_acc: 0.6643 - val_recall: 0.5690
Epoch 7/40
281/281 [==============================] - 143s 508ms/step - loss: 0.5939 - acc: 0.7006 - recall: 0.7507 - val_loss: 0.6949 - val_acc: 0.5666 - val_recall: 0.9134
Epoch 8/40
281/281 [==============================] - 143s 509ms/step - loss: 0.6102 - acc: 0.6790 - recall: 0.6997 - val_loss: 0.5610 - val_acc: 0.7114 - val_recall: 0.7454
Epoch 9/40
281/281 [==============================] - 143s 509ms/step - loss: 0.5809 - acc: 0.6986 - recall: 0.6944 - val_loss: 0.5924 - val_acc: 0.7087 - val_recall: 0.6680
Epoch 10/40
281/281 [==============================] - 143s 510ms/step - loss: 0.5860 - acc: 0.6884 - recall: 0.6699 - val_loss: 0.6226 - val_acc: 0.6679 - val_recall: 0.8791
Epoch 11/40
281/281 [==============================] - 143s 509ms/step - loss: 0.5794 - acc: 0.7151 - recall: 0.7127 - val_loss: 6.4908 - val_acc: 0.5400 - val_recall: 0.0860
Epoch 12/40
281/281 [==============================] - 143s 509ms/step - loss: 0.5779 - acc: 0.7151 - recall: 0.7210 - val_loss: 0.7116 - val_acc: 0.6750 - val_recall: 0.7996
Epoch 13/40
281/281 [==============================] - 143s 508ms/step - loss: 0.6010 - acc: 0.6846 - recall: 0.7041 - val_loss: 0.5896 - val_acc: 0.7096 - val_recall: 0.8212


I am using MacOS 10.14.5 with the Vega20 GPU. I use the your latest plaidml distribution for Mac with the following versions:
plaidml                   0.6.4                 
plaidml-keras             0.6.4       
tensorflow-gpu            1.1.0   
tensorflow                1.14.0 
keras                     2.2.4  

in my Code I use: 
import plaidml.keras
plaidml.keras.install_backend()

before importing keras

           
I would really appreciate if you had any thoughts on the matter?
